### ISSUE TITLE
what does Policy_loss mean?

### ISSUE THREAD

**Issue by MarkAlive** [Astrodyn94/SNAS-Stochastic-Neural-Architecture-Search-#2]
hi @Astrodyn94 , thanks for your code!  I do not understand the purpose of using Policy_loss in model train. Can you explian it? Is there a corresponding description in the paper? THX

**Comment by MarkAlive** [comment]
I found that if I remove this loss, the alpha para will not be updated after calling optimizer_arch.step()

**Comment by MarkAlive** [comment]
But if I maintance policy_loss, the training speed will be limited(20 samples/secend).

**Comment by zanonShao** [comment]
> hi @Astrodyn94 , thanks for your code! I do not understand the purpose of using Policy_loss in model train. Can you explian it? Is there a corresponding description in the paper? THX

hello MarkAlive, I have the same question as you, so do you know what's the purpose of using Policy_loss now? If you do, I hope you can give me some pointers. THX

**Comment by PLoic** [comment]
Hello, from what I have read so far and from what I understand, the alphas could be updated thanks to ``score_function``, have you try to just set ``policy_loss = score_function `` ?

For my part, I do not understand the meaning of the ``model.Credit``  in the equation, if it is related to the  "Credit Assignment" in the paper, why are the inputs multiplied by the grads? [#L142](https://github.com/Astrodyn94/SNAS-Stochastic-Neural-Architecture-Search-/blob/master/model_search.py#L142)
Any ideas on this part ?

**Comment by bigestboss** [comment]
> Hello, from what I have read so far and from what I understand, the alphas could be updated thanks to `score_function`, have you try to just set `policy_loss = score_function ` ?
> 
> For my part, I do not understand the meaning of the `model.Credit` in the equation, if it is related to the "Credit Assignment" in the paper, why are the inputs multiplied by the grads? [#L142](https://github.com/Astrodyn94/SNAS-Stochastic-Neural-Architecture-Search-/blob/master/model_search.py#L142)
> Any ideas on this part ?

I think `model.Credit` maybe Formula 8 in the paper. 
However, I find this code have not used 'Reparameterization' which is mentioned in the paper. So I repalce the `model.Credit` and `score_function` by  `rsample()` as:
    ```
def ArchitectDist(self,alpha,temperature):
        m = torch.distributions.relaxed_categorical.RelaxedOneHotCategorical(
            torch.tensor([temperature]).cuda() , alpha)
        return m.rsample()
```
also remove `policy_loss` and only use `value_loss`. By this method I can update alpha as mentioned in the paper.

**Comment by zanonShao** [comment]
> Hello, from what I have read so far and from what I understand, the alphas could be updated thanks to `score_function`, have you try to just set `policy_loss = score_function ` ?
> 
> For my part, I do not understand the meaning of the `model.Credit` in the equation, if it is related to the "Credit Assignment" in the paper, why are the inputs multiplied by the grads? [#L142](https://github.com/Astrodyn94/SNAS-Stochastic-Neural-Architecture-Search-/blob/master/model_search.py#L142)
> Any ideas on this part ?

I think `model.Credit` maybe Formula 8 in the paper. 
However, I find this code have not used 'Reparameterization' which is mentioned in the paper. So I repalce the `model.Credit` and `score_function` by  `rsample()` as:
```
def ArchitectDist(self,alpha,temperature):
        m = torch.distributions.relaxed_categorical.RelaxedOneHotCategorical(
            torch.tensor([temperature]).cuda() , alpha)
        return m.rsample()
```
also remove `policy_loss` and only use `value_loss`. By this method I can update alpha as mentioned in the paper.

**Comment by PLoic** [comment]
Thank you for your answer, that was also what I thought, but I doubted because the formula was different from the one in the paper.

Good point for the "Reparameterization trick", thanks :+1:

**Comment by shuida** [comment]
@MarkAlive @bigestboss @PLoic I have the same question as you. So what is the final conclusion?  Have you reproduce the same accuracy as reported in the paper?
